### PR TITLE
Certificates: DB management for Static Certificates #76

### PR DESCRIPTION
**Pull request by Pols95**
New kind of certificate: MANUAL:
* loadble only from api (.p12 file with no pw)
* stored to db but managed by admin (renewal)

ConfigurationStore:
* ability to retrive next available index for a property-name
* check whether a regexp matches with an existing property (key+value)
* get all properties

SSlCertificateConfiguration:
* property sslcertfile renamed to file
* property sslcertfilepassword renamed to password
* property iddynamic removed
* new property mode = acme|static|manual added

Dependecies:
* updated herddb to 0.10.1

Refactoring:
* extract keystore store-load-check methods to CertificateUtils class

### CODE REVIEW

File: carapace-server/src/main/java/org/carapaceproxy/api/CertificatesResource.java [hunk 1]
```diff
@@ -183,4 +199,33 @@ static String stateToStatusString(DynamicCertificateState state) {
...
+    @POST
+    @Path("{domain}/upload")
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    public Response uploadCertificate(@PathParam("domain") String domain,
+            @FormDataParam("file") InputStream uploadedInputStream) throws Exception {
+
+        HttpProxyServer server = (HttpProxyServer) context.getAttribute("server");
+        ConfigurationStore store = server.getDynamicConfigurationStore();        
+
+        // Salvo il certificato su db
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/api/CertificatesResource.java]: please leave comments in English

File: carapace-server/src/main/java/org/carapaceproxy/api/CertificatesResource.java [hunk 2]
```diff
@@ -183,4 +199,33 @@ static String stateToStatusString(DynamicCertificateState state) {
...
+        String encodedChain = Base64.getEncoder().encodeToString(uploadedInputStream.readAllBytes());
+        CertificateData cert = new CertificateData(domain, "", encodedChain, AVAILABLE.name(), "", "", true);
+        store.saveCertificate(cert);
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/api/CertificatesResource.java]: this stuff must be in HttpServer:
- save certificate
- update configuration
- apply configuration

from line 215 to line 226

File: carapace-server/src/main/java/org/carapaceproxy/configstore/ConfigurationStore.java [hunk 3]
```diff
@@ -30,22 +34,55 @@
...
     void forEach(BiConsumer<String, String> consumer);
 
-    void forEach(String prefix, BiConsumer<String, String> consumer);
+    void forEach(String prefix, BiConsumer<String, String> consumer);    
 
     default Properties asProperties(String prefix) {
         Properties copy = new Properties();
         this.forEach((k, v) -> {
-            if (k.startsWith(prefix)) {
+            if (prefix == null) {
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/configstore/ConfigurationStore.java]: this is not clear, please add a new method or add a javadoc comment

File: carapace-server/src/main/java/org/carapaceproxy/configstore/ConfigurationStore.java [hunk 4]
```diff
@@ -30,22 +34,55 @@
...
+
+            }
+        });
+        return 1 + usedIndexes.stream().max(Comparator.naturalOrder()).orElse(-1);
+    }
+
+    default boolean anyPropertyMatches(String regexp) {
+        BooleanHolder any = new BooleanHolder(false);
+        this.forEach((k, v) -> {
+            boolean matches = (k + "=" + v).matches(regexp);
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/configstore/ConfigurationStore.java]: this is not good you should use a BiPredicate

File: carapace-server/src/main/java/org/carapaceproxy/server/RuntimeServerConfiguration.java [hunk 5]
```diff
@@ -285,18 +286,24 @@ public void configure(ConfigurationStore properties) throws ConfigurationNotVali
 
     private void tryConfigureCertificate(int i, ConfigurationStore properties) throws ConfigurationNotValidException {
         String prefix = "certificate." + i + ".";
-
-        String certificateHostname = properties.getProperty(prefix + "hostname", "");
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/RuntimeServerConfiguration.java]: nit: it is not a good practice to rename variable if not needed, you are making the diff harder to understand

File: carapace-server/src/main/java/org/carapaceproxy/server/certiticates/DynamicCertificatesManager.java [hunk 6]
```diff
@@ -150,7 +150,11 @@ private void loadCertificates(Map<String, SSLCertificateConfiguration> certifica
                 SSLCertificateConfiguration config = e.getValue();
                 if (config.isDynamic()) {
                     String domain = config.getHostname();
-                    _certificates.put(domain, loadOrCreateDynamicCertificateForDomain(domain));
+                    CertificateData cert = loadOrCreateDynamicCertificateForDomain(domain);
+                    if (MANUAL.equals(config.getMode())) {
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/certiticates/DynamicCertificatesManager.java]: nit: '==' for enums

File: carapace-server/src/test/java/org/carapaceproxy/server/certiticates/DynamicCertificatesManagerTest.java [hunk 7]
```diff
@@ -71,10 +71,10 @@
 
     @Test
     @Parameters({
-        "challenge_null",
-        "challenge_status_invalid",
-        "order_response_error",
-        "available_to_expired",
+//        "challenge_null",
```
Comment (eolivelli) [inline carapace-server/src/test/java/org/carapaceproxy/server/certiticates/DynamicCertificatesManagerTest.java]: commented?

File: carapace-server/src/main/java/org/carapaceproxy/server/HttpProxyServer.java [hunk 8]
```diff
@@ -553,6 +554,22 @@ public RuntimeServerConfiguration buildValidConfiguration(ConfigurationStore sim
         return newConfiguration;
     }
 
+    public void createDynamicCertificateForDomain(CertificateData cert) throws Exception {
+        // Certificate saving on db
+        dynamicConfigurationStore.saveCertificate(cert);
+
+        // Configuration updating with new certificate (whether not yet existent)
+        Properties props = dynamicConfigurationStore.asProperties(null);
+        if (!dynamicConfigurationStore.anyPropertyMatches("certificate\\.[0-9]+\\.hostname", cert.getDomain())) {
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/HttpProxyServer.java]: cert.getDomain() -> this is not a regexp
I have suggested to use BiPredicate or BiFunction<Boolean, String, String> this way the match login will be here and not pushed into the ConfigurationStore

File: carapace-server/src/main/java/org/carapaceproxy/server/certiticates/DynamicCertificatesManager.java [hunk 9]
```diff
@@ -150,7 +150,11 @@ private void loadCertificates(Map<String, SSLCertificateConfiguration> certifica
                 SSLCertificateConfiguration config = e.getValue();
                 if (config.isDynamic()) {
                     String domain = config.getHostname();
-                    _certificates.put(domain, loadOrCreateDynamicCertificateForDomain(domain));
+                    CertificateData cert = loadOrCreateDynamicCertificateForDomain(domain);
+                    if (MANUAL == config.getMode()) {
+                        cert.setManual(true);   // in-memory only
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/certiticates/DynamicCertificatesManager.java]: the comment seems wrong "in-memory only"

File: carapace-server/src/main/java/org/carapaceproxy/api/CertificatesResource.java [hunk 10]
```diff
@@ -183,4 +202,36 @@ static String stateToStatusString(DynamicCertificateState state) {
...
+                return "unknown";
+        }
+    }
+
+    @POST
+    @Path("{domain}/upload")
+    @Consumes(MediaType.APPLICATION_OCTET_STREAM)
+    public Response uploadCertificate(@PathParam("domain") String domain, InputStream uploadedInputStream) throws Exception {
+        
+        String encodedChain = Base64.getEncoder().encodeToString(uploadedInputStream.readAllBytes());
```
Comment (Pols95) [inline carapace-server/src/main/java/org/carapaceproxy/api/CertificatesResource.java]: add a validation

**Pull request merged**